### ISSUE TITLE
Regex different result betwwen Python 2 and Python 3

### ISSUE THREAD

**Issue by Cabu** [python/cpython#109579]
# Bug report

### Bug description:

```python
import re
html_text = '<p>sentence 1</p><p>sentence 2</p>'
pattern = re.compile("(?:^|(?<=>))([\\s\\S]*?)(?:(?=<)|$)")
sentence_list = list(set(pattern.findall(html_text)))
```

In Python 2.7.18, the re module return:
```python
sentence_list = ['sentence 1', 'sentence 2']
```

In Python 3.7.16 and 3.11.3, the re module return.
```python
sentence_list = ['<p>sentence 1', '<p>sentence 2']
```

Did I miss something in the Docs (most probably)?
Regex 101 output the same result than Python 2 (https://regex101.com/r/SADcyq/1)

### CPython versions tested on:

3.11

### Operating systems tested on:

Linux

**Comment by pochmann** [comment]
From the `re.findall` [doc](https://docs.python.org/3/library/re.html#re.findall):

> Changed in version 3.7: Non-empty matches can now start just after a previous empty match.

You didn't actually get those result lists, did you? They actually contained the empty string, no?

**Comment by serhiy-storchaka** [comment]
Concur with @pochmann.

It may easier better to understand what it found if look at indices instead of substrings:
```python
[m.span() for m in pattern.finditer(html_text)]
```
In Python 3.7+ you get `[(0, 0), (0, 13), (17, 17), (17, 30), (34, 34)]` and in older versions you get `[(0, 0), (3, 13), (17, 17), (20, 30), (34, 34)]`. All versions found empty strings at positions 0, 17 and 34. But Python older than 3.7 failed to find a match started just past the previous empty match.

In short, there was a bug in old Python, it was fixed in Python 3.7, but the fix was not backported because some programs can rely on that bug.
